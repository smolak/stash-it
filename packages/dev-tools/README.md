# `@stash-it/dev-tools`

![logo-stash-it-color-dark 2x](https://user-images.githubusercontent.com/1819138/30385483-99fd209c-98a7-11e7-85e2-595791d8d894.png)

Set of tools for development purposes.

## Table of Contents

- [Installation](#installation)
- [Usage](#usage)
- [License](#license)
- [Contribution](#contribution)

npm
```bash
npm install @stash-it/dev-tools
```

deno
```bash
deno add @stash-it/dev-tools
```

yarn
```bash
yarn dlx jsr add @stash-it/dev-tools
```

pnpm
```bash
pnpm dlx jsr add @stash-it/dev-tools
```

bun
```bash
bunx jsr add @stash-it/dev-tools
```

## Usage

### `runAdapterTests`

If you want to create an adapter, it should pass all of those tests. Here's how you can use them.
Let's say you use vitest:

```typescript
// Vitest stuff
import { describe } from "vitest";

import { adapterTestCases } from '@stash-it/dev-tools';

// Your adapter.
import { YourAdapter } from './your-adapter';

describe('YourAdapter', () => {
  const adapter = new YourAdapter();
  
  // Add whatever validation or other stuff you need.
  
  // Run the tests.
  adapterTestCases(adapter);
});
```

This helper function accepts an adapter instance and runs tests on it. It will throw an error if any of the tests fail.
It is also capable of cleaning up after self. Meaning, if you're operating on a persistent storage, it will delete all of the
items created during the tests.

### `getHandler`

If you want to creat a plugin, you can use this helper function to get a handler for a specific hook.
Here's how you can use it:

```typescript
// Vitest stuff
import { describe, it, expect } from "vitest";

import { getHandler } from '@stash-it/dev-tools';

// Your plugin.
import { createPlugin } from './your-plugin';

describe('YourPlugin', () => {
  it('does something on buildKey', async () => {
    const plugin = createPlugin();
    const key = "key";
    const handler = getHandler('buildKey', plugin);
    
    // Here is where you can test the handler that you created.
    const result = await handler(key);
    
    // expect ... etc.
  });
});
```

Why not simply do this?

```typescript
const handler = plugin.hookHandlers.buildKey;
```

Well ... current types set around plugins are not that great.
This helper function will throw an error if the handler is not a function. Also, thanks to the argument passed (hook name),
it will know what arguments the handler should accept etc. And you don't need to reach to plugin internals. You simply
request a handler for a specific hook and that's it.

## License

MIT

## Contribution

Feel free to open an issue or a pull request.
