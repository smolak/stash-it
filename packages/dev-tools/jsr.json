{
  "name": "@stash-it/dev-tools",
  "version": "0.0.0-beta.4",
  "exports": "./mod.ts",
  "publish": {
    "include": ["package.json", "README.md", "mod.ts", "src/**/*.ts"],
    "exclude": ["src/**/*.test.ts"]
  }
}
