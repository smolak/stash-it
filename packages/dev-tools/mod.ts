export * from "./src/test/adapter-test-cases";
export * from "./src/test/helpers";
