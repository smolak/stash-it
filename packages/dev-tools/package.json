{
  "name": "@stash-it/dev-tools",
  "version": "0.0.5",
  "type": "module",
  "publishConfig": {
    "access": "public"
  },
  "exports": "./mod.ts",
  "files": [
    "mod.ts",
    "src",
    "_dist",
    "README.md",
    "package.json"
  ],
  "license": "MIT",
  "keywords": [
    "@stash-it",
    "stash it",
    "stash-it",
    "dev-tools",
    "tools"
  ],
  "scripts": {
    "lint": "eslint .",
    "publish": "pnpx jsr publish"
  },
  "dependencies": {
    "@stash-it/core": "npm:@jsr/stash-it__core@0.3.0",
    "nanoid": "^5.0.8",
    "vitest": "^2.1.4"
  }
}
