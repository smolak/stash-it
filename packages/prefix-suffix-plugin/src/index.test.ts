import { describe, it, expect } from "vitest";
import { getHandler } from "@stash-it/dev-tools";

import { createPrefixSuffixPlugin } from "./index";

describe("prefix-suffix-plugin", () => {
  describe("when prefix is set", () => {
    it("builds key with set prefix", async () => {
      const plugin = createPrefixSuffixPlugin({ prefix: "prefix-" });
      const key = "key";
      const handler = getHandler("buildKey", plugin);

      const result = await handler({ key });

      expect(result.key).toEqual(`prefix-${key}`);
    });
  });

  describe("when suffix is set", () => {
    it("builds key with set suffix", async () => {
      const plugin = createPrefixSuffixPlugin({ suffix: "-suffix" });
      const key = "key";
      const handler = getHandler("buildKey", plugin);

      const result = await handler({ key });

      expect(result.key).toEqual(`${key}-suffix`);
    });
  });

  describe("when prefix and suffix are set", () => {
    it("builds key with set prefix and suffix", async () => {
      const plugin = createPrefixSuffixPlugin({ prefix: "prefix-", suffix: "-suffix" });
      const key = "key";
      const handler = getHandler("buildKey", plugin);

      const result = await handler({ key });

      expect(result.key).toEqual(`prefix-${key}-suffix`);
    });
  });

  describe("when neither prefix nor suffix is set", () => {
    it("throws an error", () => {
      expect(() => createPrefixSuffixPlugin({})).toThrowErrorMatchingSnapshot();
    });
  });
});
