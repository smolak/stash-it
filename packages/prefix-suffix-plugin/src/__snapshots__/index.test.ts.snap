// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`prefix-suffix-plugin > when neither prefix nor suffix is set > throws an error 1`] = `
[ZodError: [
  {
    "message": "Either prefix or suffix should be set.",
    "code": "custom",
    "path": [
      "prefix"
    ]
  },
  {
    "message": "Either prefix or suffix should be set.",
    "code": "custom",
    "path": [
      "suffix"
    ]
  }
]]
`;
